## Snakefile pipeline for the phasing common and rare variants with SHAPEIT5

The working directory will be the one from where this script will be executed
The pipeline consists of 4 small steps (conceptually, might be tough computationally)
    1) phase common for the whole cohort
    2) phase rare for the whole cohort, in chunks
    3) phase trios (both common and rare variants)
    4) assess phasing and get files with the PP distribution

#### Required input: ####
1. Prepared BCF files, per chromosome, for {trios,notrios} x {common,rare} (ie 4x22 files). 
    These can be generated by `smk_01_prep_all_bcf.sh`. See that file for more. 
2. {TAG}.100trios.pedigree: a list of {child,father,mother} triplets for trio phasing
3. Genetic maps and lists of chunks, per chromosome (see the corresponding scripts)
4. A few parameters for Shapeit5 which are already set

### Important notes ####
* [TODO] File prep is not included here as is tricky; this could be performed with `smk_00_merge_wes_array.sh`
* Each script contains a few params, such as paths to genetic maps, or chunk lists.
* Software dependencies: bcftools (+HTSlib), SHAPEIT5, snakemake (+Python); 
* I'm currently using a combination of binaries and modules for Shapeit5 (to deal with issues 34,56), please choose what's best
* [TODO] The SER analysis is performed using the `switch` tool provided by SHAPEIT5 - this might need to change , as we can similarly work with `bcftools +trio-switch-rate ...`, which requires simpler input.
* [TODO] use ligate instead of concat within phase-rare

#### How to - LSF ####
* the following will start the pipeline and submit all jobs with all the jobs required:
`snakemake --cluster "bsub -M 16G -R 'select[mem>16G] rusage[mem=16G]' -n10 -G team281 -q normal -o run_all.stdout -e run_all.stderr" --jobs 10 --latency-wait 10 -T 2 run_all`
* first run the following the initialise the required folders:
`mkdir -p phased_genotypes_common phased_genotypes_rare phasing_assessment sandbox`

GK - Sep 26th, 2023